tests/console/zfs.pm

```diff
@@ -0,0 +1,190 @@
+    } else {
+        die "Unsupported version";
+    }
+    zypper_call('refresh');
+    zypper_call('install zfs');
+}
+
+sub prepare_disks {
+    assert_script_run('fallocate -l 1GB /var/tmp/tank_a.img');
+    assert_script_run('fallocate -l 1GB /var/tmp/tank_b.img');
+    assert_script_run('fallocate -l 1GB /var/tmp/tank_c.img');
+    assert_script_run('fallocate -l 1GB /var/tmp/dozer.img');
+}
+
+sub corrupt_disk {
+    my $disk = shift;
+    assert_script_run("dd if=/dev/zero of=$disk bs=1G count=1");
+}
+
+sub cleanup {
+    script_run('cd');
+    script_run('if zpool list | grep tank ; then zpool destroy tank; fi');
+    script_run('if zpool list | grep dozer ; then zpool destroy dozer; fi');
+    script_run('rm -f /var/tmp/tank_{a,b,c}.img /var/tmp/dozer.img');
+}
+
+sub scrub {
+    my $pool = shift;
+    assert_script_run("zpool scrub $pool");
+    # Wait for scrub to finish
+    script_retry("zpool status $pool | grep scan | grep -v 'in progress'", delay => 1, retry => 30);
+}
+
+sub reboot {
+    my ($self) = @_;
+    power_action('reboot', textmode => 1);
+    $self->wait_boot(bootloader_time => 300);
+    select_console 'root-console';
+}
+
+sub run {
+    # Preparation
+    my $self = shift;
+    # TODO: Switch to serial_terminal for BLAZING FAST EXECUTING!!!
+    #$self->select_serial_terminal;
+    select_console 'root-console';
+    install_zfs();
+    assert_script_run('modprobe zfs');
+    prepare_disks();
+    # Create zfs test pools
+    assert_script_run('zpool create tank raidz /var/tmp/tank_{a,b,c}.img -o ashift=13');
+    assert_script_run('zpool list | grep tank');
+    assert_script_run('zpool create dozer /var/tmp/dozer.img -o ashift=12');
+    assert_script_run('zpool list | grep dozer');
+    assert_script_run('cd /tank/; ls');
+    # Put some data on tank
+    assert_script_run('curl -v -o /tank/Big_Buck_Bunny_8_seconds_bird_clip.ogv ' . data_url('Big_Buck_Bunny_8_seconds_bird_clip.ogv'));
+    assert_script_run('md5sum Big_Buck_Bunny_8_seconds_bird_clip.ogv > /var/tmp/Big_Buck_Bunny_8_seconds_bird_clip.ogv.md5sum');
+    assert_script_run('md5sum -c /var/tmp/Big_Buck_Bunny_8_seconds_bird_clip.ogv.md5sum');
+    scrub('tank');
+    assert_script_run('zpool status tank | grep state | grep ONLINE');
+    script_run('zpool status tank');
+    save_screenshot();
+    ## Test raidz capability
+    corrupt_disk('/var/tmp/tank_a.img');
+    # Wait until degraded
+    assert_script_run('md5sum -c /var/tmp/Big_Buck_Bunny_8_seconds_bird_clip.ogv.md5sum');
+    scrub('tank');
+    assert_script_run("zpool status tank | grep scan | grep 'scrub repaired'");
+    assert_script_run('md5sum -c /var/tmp/Big_Buck_Bunny_8_seconds_bird_clip.ogv.md5sum');
+    script_run('zpool status tank');
+    assert_script_run('zpool status tank | grep "corrupted data"');
+    # Replace disk
+    assert_script_run('zpool offline tank /var/tmp/tank_a.img');
+    assert_script_run('zpool replace tank /var/tmp/tank_a.img');
```

Comment:
Is it OK that you replaced one disk by the same so you actually didn't replaced it but just replugged it?

Reply:
`zpool replace tank DISK` replaces the affected disk with a new disk in the same location, so it should be fine. It's similar to unplugging a hot-swappable disk and replace it with a new one.

See [Replacing a Device in a ZFS Storage Pool](https://docs.oracle.com/cd/E19253-01/819-5461/gbcet/index.html)

IMHO this should be fine.

Reply:
No, you are replacing `/var/tmp/tank_a.img` by `/var/tmp/tank_a.img` or am I lost somewhere else? 😆

Reply:
`zpool replace /var/tmp/tank_a.img` **marks** this disk as being replaced and thus should be fine.

The procedure is the following:

    assert_script_run('zpool offline tank /var/tmp/tank_a.img');
    clear_disk('/var/tmp/tank_a.img');
    assert_script_run('zpool replace tank /var/tmp/tank_a.img');

This is the same as taking the disk offline, replace it and (somehow it ends up in the same location) then mark the disk as replaced. Think of `clear_disk` as "I take the disk out and hot-swap it with a new one". Think of `zpool replace` as "disk is marked as replaced"

Reply:
Ah yes, I missed the `clear_disk('/var/tmp/tank_a.img');` step. Sorry Felix!

Reply:
You didn't miss it, I added it afterwards ;-)